### PR TITLE
[FIX] account: add constrains on payment method line journal

### PR DESCRIPTION
**Pull request by Arnaud-Sibille**
When you create or register a payment, the domain of the payment method lines is based on a computed field dependent of the selected journal. The issue is that the interface isn't blocked when waiting for the onchange return.  So if the user changes the journal, then select the payment method quickly enough while the onchange is still pending, he will be able to select outdated values.

It is a limitation of the js framework, so to avoid the user encoding wrong datas, the fix here is to raise a `ValidationError` telling to re-select the payment method.

To reproduce:
- create second bank journal, with outbound payment method lines having different names than the ones of the first bank journal (in order to distinct them).
- slow down the `_compute_payment_method_line_fields` method
- create a vendor payment, switch the journal to the one created and select the second payment method (before the onchange ends).
- save the payment.

-> The payment has a payment method line from a different journal

opw-3587241

Enterprise PR: https://github.com/odoo/enterprise/pull/56602

### CODE REVIEW

File: addons/account/models/account_payment.py [hunk 1]
```diff
@@ -638,6 +638,12 @@ def _check_payment_method_line_id(self):
             if not pay.payment_method_line_id:
                 raise ValidationError(_("Please define a payment method line on your payment."))
 
+    @api.constrains('payment_method_line_id', 'journal_id')
+    def _check_payment_method_line_journal(self):
+        for payment in self:
+            if payment.payment_method_line_id.journal_id and payment.payment_method_line_id.journal_id != payment.journal_id:
+                raise ValidationError("The selected payment method is from a different journal, please select the payment method again.")
```
Comment (aliyatastemirova) [inline addons/account/models/account_payment.py:645]: this should be translated. Do we really need to duplicate this constraint in the wizard?

File: addons/account/tests/test_reconciliation.py [hunk 2]
```diff
@@ -271,7 +271,7 @@ def test_partial_reconcile_currencies_01(self):
             'company_id': self.company.id,
             'date': time.strftime('%Y') + '-' + '07' + '-01',
             'partner_id': self.partner_agrolait_id,
-            'payment_method_line_id': self.inbound_payment_method_line.id,
+            'payment_method_line_id': self.bank_journal_euro.outbound_payment_method_line_ids[0].id,
```
Comment (aliyatastemirova) [inline addons/account/tests/test_reconciliation.py:274]: can you please clarify why this change is needed? thanks

**Review by aliyatastemirova — CHANGES_REQUESTED** [review]
hi, thanks :slightly_smiling_face: just some small comments

File: addons/account/models/account_payment.py (hunk 1 above)
Comment (aliyatastemirova) [reply addons/account/models/account_payment.py:645]: also, I think we should check `available_payment_method_lines` here, because even if the journal is the same, the payment method might still be invalid (outbound/inbound/filtered/etc)

File: addons/account/tests/test_reconciliation.py (hunk 2 above)
Comment (Arnaud-Sibille) [reply addons/account/tests/test_reconciliation.py:274]: There are tests that are failing due to the constrains

Here for example the payment journal is `self.bank_journal_id`, so a payment method line of that journal has to be set

File: addons/account/models/account_payment.py (hunk 1 above)
Comment (Arnaud-Sibille) [reply addons/account/models/account_payment.py]: I've adapted the changes :)

Just have 2 questions:
Should I add the constrains of `available_payment_method_line_ids` (`journal_id`, `payment_type`, `currency_id`)?

There's something that will fail now but don't know if that matters:
It is possible to remove a payment method line from a journal, and so the payment method line does not have any `journal_id` anymore. 
So if you reset a to draft a payment that had that method line and re-post it, the validation error will be raised as this method line is no longer in the available ones.
Don't know if that matters, the user can still select a new one.

Comment (Arnaud-Sibille) [reply addons/account/models/account_payment.py]: > Do we really need to duplicate this constraint in the wizard?

No, don't know why I did that

Comment (aliyatastemirova) [reply addons/account/models/account_payment.py]: do you mean that if you reset a payment to draft, then remove that payment method from the journal, and then try to confirm the payment, the error will be raised? If so, in my opinion, that's ok.
there are more tests failing though :slightly_smiling_face:

Comment (aliyatastemirova) [reply addons/account/models/account_payment.py]: oh, and the translation should be added in the .pot file :slightly_smiling_face:

File: addons/account/tests/test_account_payment_register.py [hunk 3]
```diff
@@ -320,12 +320,12 @@ def test_register_payment_single_batch_grouped_writeoff_higher_amount_credit(sel
             'payment_difference_handling': 'reconcile',
             'writeoff_account_id': self.company_data['default_account_revenue'].id,
             'writeoff_label': 'writeoff',
-            'payment_method_line_id': self.inbound_payment_method_line.id,
+            'payment_method_line_id': self.outbound_payment_method_line.id,
         })._create_payments()
 
         self.assertRecordValues(payments, [{
             'ref': 'BILL/2017/01/0001 BILL/2017/01/0002',
-            'payment_method_line_id': self.inbound_payment_method_line.id,
```
Comment (aliyatastemirova) [inline addons/account/tests/test_account_payment_register.py:328]: if it's a vendor bill, shouldn't it be an inbound payment method? :thinking:

Comment (Arnaud-Sibille) [reply addons/account/tests/test_account_payment_register.py:328]: I don't think so, you have to pay your bill, so the payment is 'outbound' and so the payment method

File: addons/account/models/account_payment.py (hunk 1 above)
Comment (Arnaud-Sibille) [reply addons/account/models/account_payment.py]: > there are more tests failing though 🙂

The last 2 tests that fails are for payment transactions refunds.
I wonder if it is not a bug in the code:
Here we see that the payment can be inbound/outbound https://github.com/odoo/odoo/blob/799465ec53e0a07890ae6bcf3b0b7b5b46fdc00b/addons/payment/models/payment_transaction.py#L930
But it will only take inbound payment methods: https://github.com/odoo/odoo/blob/799465ec53e0a07890ae6bcf3b0b7b5b46fdc00b/addons/payment/models/payment_transaction.py#L926

Comment (aliyatastemirova) [reply addons/account/models/account_payment.py]: > The last 2 tests that fails are for payment transactions refunds. I wonder if it is not a bug in the code: Here we see that the payment can be inbound/outbound

hmm it looks like a bug indeed :thinking:

Comment (aliyatastemirova) [reply addons/account/models/account_payment.py]: I wonder if we could just add a domain on the payment method?

Comment (aliyatastemirova) [reply addons/account/models/account_payment.py]: hi, could you please let me know if you have an update on this? :slightly_smiling_face:

File: addons/account/models/account_payment.py [hunk 4]
```diff
@@ -637,6 +637,8 @@ def _check_payment_method_line_id(self):
         for pay in self:
             if not pay.payment_method_line_id:
                 raise ValidationError(_("Please define a payment method line on your payment."))
+            elif pay.payment_method_line_id not in pay.available_payment_method_line_ids:
```
[the remark this answers is not in the document]
Comment (Arnaud-Sibille) [reply addons/account/models/account_payment.py:640]: Hello :)

This is what I originally did, it would require less tests to be changed, but some will still need to be.
Like the tests in addons/account/tests/test_reconciliation.py (they use a different journal but still use the default payment method of the setup class)

**Comment by aliyatastemirova** [comment]
@jbw-odoo yeah I think it's ok :thinking: I thought it would be an issue if the payment method was invalid but I guess it's not

**Review by aliyatastemirova — APPROVED** [review]

**Comment by aliyatastemirova** [comment]
yes, I suppose you're right :+1:

**Comment by aliyatastemirova** [comment]
...I guess it's an issue after all :joy:

**Comment by Arnaud-Sibille** [comment]
Some tests in enterprise gives `account.payment.method` record instead of `account.payment.method.line` for the `payment_method_line_id` field
Guess I will need to make a PR in enterprise

**Comment by Arnaud-Sibille** [comment]
should be okay now!
